### PR TITLE
Allow parallel access to klib zip filesystem

### PR DESCRIPTION
**Pull request by alexander-gorshenev**

### CODE REVIEW

File: compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt [hunk 1]
```diff
@@ -46,7 +53,13 @@ fun <T> File.withZipFileSystem(mutable: Boolean = false, action: (FileSystem) ->
     return try {
         action(zipFileSystem)
     } finally {
-        zipFileSystem.close()
+        val counter = counters[zipFileSystem]!!
+        if (counter == 0) {
+            zipFileSystem.close()
+            counters.remove(zipFileSystem)
+        } else {
+            counters.computeIfPresent(zipFileSystem, { t, u -> u + 1})
```
Comment (homuroll) [inline compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: `u - 1`

File: compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt [hunk 2]
```diff
@@ -46,7 +53,13 @@ fun <T> File.withZipFileSystem(mutable: Boolean = false, action: (FileSystem) ->
     return try {
         action(zipFileSystem)
     } finally {
-        zipFileSystem.close()
+        val counter = counters[zipFileSystem]!!
+        if (counter == 0) {
```
Comment (homuroll) [inline compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: Why 0? We're decrementing after this check, so probably here we need to compare with 1.

Comment (alexander-gorshenev) [reply compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: Uh, seems to be a stale cherry-pick. We've already fixed it all somewhere. Let me force push.

Comment (alexander-gorshenev) [reply compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: There you go.

File: compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt [hunk 3]
```diff
@@ -7,17 +7,24 @@ package org.jetbrains.kotlin.konan.file
...
     val attributes = hashMapOf("create" to mutable.toString())
     return try {
-        FileSystems.newFileSystem(zipUri, attributes, null)
+        FileSystems.newFileSystem(zipUri, attributes, null).also {
+            counters[it] = 1
+        }
     } catch (e: FileSystemAlreadyExistsException) {
-        FileSystems.getFileSystem(zipUri)
+        FileSystems.getFileSystem(zipUri).also {
+            counters.computeIfPresent(it, { t, u -> u + 1})
```
Comment (olonho) [inline compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: _, value -> value + 1

File: compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt [hunk 4]
```diff
@@ -46,7 +53,13 @@ fun <T> File.withZipFileSystem(mutable: Boolean = false, action: (FileSystem) ->
     return try {
         action(zipFileSystem)
     } finally {
-        zipFileSystem.close()
+        val counter = counters[zipFileSystem]!!
+        if (counter == 1) {
+            zipFileSystem.close()
+            counters.remove(zipFileSystem)
+        } else {
+            counters.computeIfPresent(zipFileSystem, { t, u -> u - 1})
```
Comment (olonho) [inline compiler/util-io/src/org/jetbrains/kotlin/konan/file/ZipUtil.kt]: `_, value -> value - 1`

**Pull request merged**